

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>epitopefinder_getepitopes.py &mdash; epitopefinder 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="epitopefinder 0.1 documentation" href="index.html" />
    <link rel="up" title="Main scripts" href="mainscripts.html" />
    <link rel="next" title="Analysis scripts" href="analysisscripts.html" />
    <link rel="prev" title="Main scripts" href="mainscripts.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="analysisscripts.html" title="Analysis scripts"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="mainscripts.html" title="Main scripts"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">epitopefinder 0.1 documentation</a> &raquo;</li>
          <li><a href="mainscripts.html" accesskey="U">Main scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="epitopefinder-getepitopes-py">
<span id="id1"></span><h1>epitopefinder_getepitopes.py<a class="headerlink" href="#epitopefinder-getepitopes-py" title="Permalink to this headline">¶</a></h1>
<p>This is a main script of the <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> package.</p>
<p>This script answers the question: How many epitopes of a specified type map to each position in a protein&#8217;s sequence? A user created input file specifies parameters used to filter the epitopes, map them to a protein sequence, and determine whether different mapped epitopes are unique.</p>
<p>The script begins by reading collection of epitopes from a CSV file <em>iedbfile</em> downloaded from the <a class="reference external" href="http://www.iedb.org">Immune Epitope Database</a>, and then classifying the epitopes as best as possible by MHC class and MHC gene. MHC class II epitopes are also classified by supertype using the classification system in <em>supertypesfile</em>.</p>
<p>The epitopes are then further filtered by MHC class (according to the <em>mhcclass</em> parameter) and by epitopelength (according to the <em>epitopelength</em> parameter) to give a set of epitopes under consideration. We then attempt to align these epitopes to the target protein sequence (or collection aligned homologous target protein sequences) specified in <em>targetprotsfile</em>. Epitopes that align to at least one of these target protein sequences with no gaps and &lt;= <em>maxmismatches</em> mismatches are considered to match the target protein(s). The alignment is performed using the MUSCLE program in <em>musclepath</em>. Finally, potentially redundant epitopes are removed according to the options set by <em>purgeredundant</em> and <em>purgeredundantoverlap</em>.</p>
<p>The results are written to the output CSV (comma-separated value) files <em>epitopelistfile</em> and <em>epitopesbysitefile</em>.</p>
<div class="section" id="running-the-script">
<h2>Running the script<a class="headerlink" href="#running-the-script" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="#epitopefinder-getepitopes-py"><em>epitopefinder_getepitopes.py</em></a> takes as input the name of a single file, the format of which is detailed below. If you have installed the package so that the scripts are the search path, you can run this script directly from the command line. For example, if you called your input file <tt class="docutils literal"><span class="pre">infile.txt</span></tt> then run:</p>
<div class="highlight-python"><pre>epitopefinder_getepitopes.py infile.txt</pre>
</div>
<p>If the script is not executable on your platform, then run:</p>
<div class="highlight-python"><pre>python epitopefinder_getepitopes.py infile.txt</pre>
</div>
<p>This will create the various output files described below.</p>
</div>
<div class="section" id="input-file-format">
<h2>Input file format<a class="headerlink" href="#input-file-format" title="Permalink to this headline">¶</a></h2>
<p>The input file should contain a series of keys followed by their values. Lines beginning with # or empty lines are ignored. The input file must have the following keys and their values:</p>
<blockquote>
<div><blockquote>
<div><ul>
<li><p class="first"><em>iedbfile</em> is a file containing epitopes in the format of downloads from the <a class="reference external" href="http://www.iedb.org">Immune Epitope Database</a>. It is important that you download only epitopes for which the assay result is positive, otherwise an exception will be raised. These files are downloadable from this database after you do a search for some specific class of epitopes. You will be given an option to download a <em>full</em> or <em>compact</em> file with the epitopes. This script is only confirmed to work with the <em>compact</em> formats. It is possible that it will break if the format of these downloads changes; however, it works with the download format as of April-15-2013. The files themselves are CSV (comma-separated values) files with an initial header line.</p>
</li>
<li><p class="first"><em>supertypesfile</em> is a file that gives the supertype classification of HLA-A and HLA-B alleles. The suggested classification scheme is that given in <a class="reference external" href="http://www.biomedcentral.com/1471-2172/9/1">BMC Immunology, 2008, 9(1)</a>, although you may want to update this if the scheme is updated further. A file containing the classification scheme from this reference is contained in some of the example analyses in the <tt class="docutils literal"><span class="pre">examples</span></tt> directories in this package. This file should contains lines that list the alleles at the <tt class="docutils literal"><span class="pre">gene*group:protein</span></tt> level of detail, followed by the supertype, as in:</p>
<div class="highlight-python"><pre>A*01:01 A01
A*01:02 Unclassified
A*68:15 A02
B*07:43 B07
B*08:01 B08</pre>
</div>
</li>
<li><p class="first"><em>mhcclass</em> specifies the class of MHC epitopes for which we are searching. It can be either I or II depending on whether we searching for class I or class II epitopes.</p>
</li>
<li><p class="first"><em>epitopelength</em> specifies the length of epitopes for which we include. It should be a list of two numbers, first the minimum length and then the maximum length. We only retain and consider epitopes with lengths &gt;= the minimum length and &lt;= the maximum length.</p>
</li>
<li><p class="first"><em>targetprotsfile</em> specifies a FASTA file containing the target protein sequence(s) to which we are mapping the epitope. It should contain one or more sequences. If there are multiple sequences, they should be aligned. The reason that you might want to include multiple homologous sequences is if the epitopes were defined on a set of closely homologous sequences. For instance, epitopes in influenza might be determined for several closely related sequences, and you might want to find epitopes that closely match any of these sequences. If there is more than one sequence in <em>targetprotsfile</em>, then these sequences should be aligned. The sequence positions for the epitope alignments are then numbered according to consecutive numbering of the first sequence. Here is an example <em>targetprotsfile</em>:</p>
<div class="highlight-python"><pre>&gt;A/Aichi/2/1968_NP
MAS-GTKRSYEQMETDGERQNATEIRASVGKMIDGIGRFYIQMCTELKLSDY
&gt;A/Brisbane/10/2007_NP
MASQGTKRSYEQMETDGDRQNATEIRASVGKMIDGIGRFYIQMCTELKLSDH</pre>
</div>
</li>
</ul>
<p>In this example, the residues are numbered sequentially according to the first sequence, so that the M is 1, the A is 2, the S is 3, the G is 5 (since 4 is a gap), etc.</p>
<ul>
<li><p class="first"><em>maxmismatches</em> specifies the stringency of the epitope alignment with the target protein sequences in <em>targetprotsfile</em>. We only retain epitopes that align to at least one of the sequences in <em>targetprotsfile</em> with &lt;= <em>maxmismatches</em>.</p>
</li>
<li><p class="first"><em>musclepath</em> specifies the path to a directory that contains the <a class="reference external" href="http://www.drive5.com/muscle/">MUSCLE</a> alignment program in the form of an executable named <tt class="docutils literal"><span class="pre">muscle</span></tt>. This script has been tested with <a class="reference external" href="http://www.drive5.com/muscle/">MUSCLE</a> version 3.8, but should also work with other similar versions. This alignment program is used to align the epitopes to the target proteins in <em>targetprotsfile</em>.</p>
</li>
<li><p class="first"><em>purgeredundant</em> specifies how we purge redundant epitopes (i.e. epitopes listed in <em>iedbfile</em> that are likely to actually refer to the same epitope). We find epitopes that overlap by &gt;= <em>purgeredundantoverlap</em> sites in the position to which they align to the target proteins. Depending on the option set for <em>purgeredundant</em>, such overlapping epitopes are removed. The epitope that is retained among overlapping epitopes is the one with the shortest epitope sequence, and the purging moves through epitopes from shortest to longest (ensuring that distinct short epitopes that overlap with the same longer epitope are not considered redundant). Possibilities for <em>purgeredundant</em> are:</p>
<blockquote>
<div><ul class="simple">
<li><em>None</em> specifies that no purging of redundant epitopes is done.</li>
<li><em>MHCgene</em> specifies that we retain overlapping epitopes if they differ in their MHC gene. So for example two overlapping epitopes that are both HLA-A would have one purged, but two overlapping epitopes that are HLA-A and HLA-B would not have one purged.</li>
<li><em>MHCsupertype</em> is an allowable option only if <em>MHCclass</em> was set to I. In this case, we retain overlapping epitopes if they differ in their MHC gene or if they differ in their MHC supertype.</li>
<li><em>MHCgroup</em> specifies that we retain overlapping epitopes if they differ in their MHC gene, MHC supertype (if class I), or MHC allele group. Recall that according to <a class="reference external" href="http://hla.alleles.org/nomenclature/naming.html">HLA nomenclature</a>, the allele group is the classification after the gene, so for example HLA-B*35:01 has a group of 35.</li>
</ul>
</div></blockquote>
</li>
</ul>
<blockquote>
<div>If two epitopes overlap and one does not have the lowest level of MHC gene / supertype / group set, then it is not retained as unique. So for example, if the same epitope is classified as HLA-B and HLA-B*35:01 and <em>purgeredundant</em> is set to <em>MHCgroup</em>, then one of the epitopes is purged because there is not enough information to ensure that they do not overlap in group.</div></blockquote>
<ul class="simple">
<li><em>purgeredundantoverlap</em> specifies the amount of overlap two epitopes must possess in the positions in which they align to the target gene before they are considered potentially redundant, and candidates for removal according to the options specified by <em>purgeredundant</em>. It should be a number, and epitopes with overlap &gt;= this number and candidates for removal.</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><em>epitopelistfile</em> is the name of an output CSV file created by this script. It lists each of the unique epitopes identified, along with the sites where it aligns, its allele, its references, and other redundant epitopes. If this file already exists, it is overwritten.</li>
<li><em>epitopesbysitefile</em> is the name of an output CSV file created by this script. For each site in the target protein, reports the number of unique epitopes that span this site. If this file already exits, it is overwritten.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="example-input-file">
<h2>Example input file<a class="headerlink" href="#example-input-file" title="Permalink to this headline">¶</a></h2>
<p>Here is an example input file for MHC class I epitopes:</p>
<div class="highlight-python"><pre># input file for epitopefinder_getepitopes.py
iedbfile Grant2013_plus_IEDB_Influenza_Tcell_compact_2013-04-15.csv
supertypesfile supertype_classification.txt
mhcclass I
epitopelength 8 12
targetprotsfile targetprotsfile.fasta
maxmismatches 1
musclepath /Users/jbloom/muscle3.8/
purgeredundant MHCsupertype
purgeredundantoverlap 8
epitopelistfile epitopeslist.csv
epitopesbysitefile epitopesbysite.csv</pre>
</div>
</div>
<div class="section" id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>This script creates the two output CSV files <em>epitopeslistfile</em> and <em>epitopesbysitefile</em>.</p>
<p><em>epitopeslistfile</em> lists each unique epitope along with its redundant counterparts. Here is an example of the first two lines:</p>
<div class="highlight-python"><pre>unique_epitope_sequence,unique_epitope_position,unique_epitope_allele,unique_epitope_information,unique_epitope_reference,redundant-1_epitope_sequence,redundant-1_epitope_position,redundant-1_epitope_allele,redundant-1_epitope_information,redundant-1_epitope_reference
"GERQNATEI","17-25","allele HLA-B*18:01; MHC class I; MHC gene B; MHC supertype B44; MHC group 18","host = Homo sapiens; source = Influenza A virus; molecule = nucleoprotein; reported position = (17, 25); assay = ELISPOT / cytokine release IFNg","Jeff Alexander; Pamuk Bilsel; Marie-France del Guercio; Aleksandra Marinkovic-Petrovic; Scott Southwood; Stephani Stewart; Glenn Ishioka; Maya F Kotturi; Jason Botten; John Sidney; Mark Newman; Alessandro Sette. Hum Immunol. 2010. PMID 20156506",,,,,</pre>
</div>
<p><em>epitopesbysitefile</em> lists all of the sites as they are numbered in sequential 1, 2, ... numbering of the aligned target proteins in <em>targetprotsfile</em>, followed by the number of unique epitopes containing that site. Here is an example of the first few lines:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">Site</span><span class="p">,</span><span class="n">NumberUniqueEpitopes</span>
<span class="mi">1</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">2</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">3</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">4</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">5</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">6</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">7</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">8</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">9</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">10</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">11</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">12</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">13</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">14</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">15</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">16</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">17</span><span class="p">,</span><span class="mi">3</span>
<span class="mi">18</span><span class="p">,</span><span class="mi">3</span>
<span class="mi">19</span><span class="p">,</span><span class="mi">3</span>
<span class="mi">20</span><span class="p">,</span><span class="mi">3</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">epitopefinder_getepitopes.py</a><ul>
<li><a class="reference internal" href="#running-the-script">Running the script</a></li>
<li><a class="reference internal" href="#input-file-format">Input file format</a></li>
<li><a class="reference internal" href="#example-input-file">Example input file</a></li>
<li><a class="reference internal" href="#output-files">Output files</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="mainscripts.html"
                        title="previous chapter">Main scripts</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="analysisscripts.html"
                        title="next chapter">Analysis scripts</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/epitopefinder_getepitopes.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="analysisscripts.html" title="Analysis scripts"
             >next</a> |</li>
        <li class="right" >
          <a href="mainscripts.html" title="Main scripts"
             >previous</a> |</li>
        <li><a href="index.html">epitopefinder 0.1 documentation</a> &raquo;</li>
          <li><a href="mainscripts.html" >Main scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Jesse D. Bloom.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>