<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>CTL epitopes in influenza NP: human H3N2 and swine &mdash; epitopefinder 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="epitopefinder 0.1 documentation" href="index.html" />
    <link rel="up" title="Examples" href="examples.html" />
    <link rel="next" title="Python modules" href="pythonapi.html" />
    <link rel="prev" title="Examples" href="examples.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pythonapi.html" title="Python modules"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="examples.html" title="Examples"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">epitopefinder 0.1 documentation</a> &raquo;</li>
          <li><a href="examples.html" accesskey="U">Examples</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="ctl-epitopes-in-influenza-np-human-h3n2-and-swine">
<h1>CTL epitopes in influenza NP: human H3N2 and swine<a class="headerlink" href="#ctl-epitopes-in-influenza-np-human-h3n2-and-swine" title="Permalink to this headline">¶</a></h1>
<p>This is the analysis described in <a class="reference external" href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1004328">Gong and Bloom, PLoS Genetics, 2014</a>.</p>
<p>This is an analysis of human CTL epitopes in the nucleoprotein (NP) of human H3N2 (1968 to 2013) and the swine NPs descended from the 1918 virus. It finds MHC class I epitopes with experimentally verified CTL responses as mined from the <a class="reference external" href="http://www.iedb.org">Immune Epitope Database</a>. The analysis examines the association between epitopes and sites that have substituted during the natural evolution, and sites that have been experimentally identified as being involved in epistatic interactions.</p>
<p>This analysis utilizes the <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> package. The data and scripts can be found in the subdirectory <tt class="docutils literal"><span class="pre">./examples/NP_CTL_epitopes_H3N2_and_swine</span></tt> in the <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> package.</p>
<p>This analysis was performed by <a class="reference external" href="http://research.fhcrc.org/bloom/en.html">Jesse Bloom</a>.</p>
<div class="section" id="software-used">
<h2>Software used<a class="headerlink" href="#software-used" title="Permalink to this headline">¶</a></h2>
<p>This analysis was performed on Mac OS X (10.6.8) using the following software packages:</p>
<ul class="simple">
<li><a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> (version 0.1) is the primary software package used.</li>
<li><a class="reference external" href="http://www.python.org/">Python</a> (version 2.6.8) is used to run <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a>.</li>
<li><a class="reference external" href="http://matplotlib.org/">matplotlib</a> (version 1.2.1) is used by <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> to make plots.</li>
<li><a class="reference external" href="http://www.scipy.org/">scipy</a> (version 0.12.0) is used to compute statistics.</li>
<li><a class="reference external" href="http://www.drive5.com/muscle/">MUSCLE</a> (version 3.8) is used to perform alignments.</li>
<li>The <a class="reference external" href="http://www.imagemagick.org/script/convert.php">ImageMagick convert</a> (version 6.8.5-5) command-line tool is used to create JPG versions of the plots created by <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a></li>
<li>The <a class="reference external" href="http://linux.about.com/library/cmd/blcmdl1_pdf2ps.htm">pdf2ps</a> command-line tool (from Ghostscript 9.06) is used to create PS versions of the plots created by <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a>.</li>
<li>The <a class="reference external" href="http://www.ctan.org/pkg/ps2eps">ps2eps</a> (version 1.64) command-line tool is used to create EPS files of the plots from the PS files.</li>
</ul>
</div>
<div class="section" id="input-data-files">
<h2>Input data files<a class="headerlink" href="#input-data-files" title="Permalink to this headline">¶</a></h2>
<p>The following input files were used for this analysis:</p>
<blockquote>
<div><ul class="simple">
<li><em>IEDB_Influenza_Tcell_compact_2013-12-27.csv</em> is a CSV file containing the compact format of the downloaded epitopes for all assays in the <a class="reference external" href="http://www.iedb.org">Immune Epitope Database</a> with positive T-cell responses with the source organism being &#8220;Influenza A virus&#8221; and the host being &#8220;Homo sapiens&#8221; as downloaded on December-27-2013.</li>
<li><em>supertype_classification.txt</em> is a text file that classifies the HLA-A and HLA-B alleles according to supertypes. This supertype classification scheme was taken from <a class="reference external" href="http://www.biomedcentral.com/1471-2172/9/1">BMC Immunology, 2008, 9(1)</a> by downloading Additional file 1 from this reference, and then cutting out the supertype classification columns and adding a : character between the group and protein in the allele names.</li>
<li><em>targetprots_humanH3N2.fasta</em> contains the aligned sequences of the NP proteins from the human H3N2 strains A/Aichi/2/1968 and A/Texas/JMM_49/2012.</li>
<li><em>targetprots_swine.fasta</em> contains the aligned sequences of the NP proteins from the swine 1918-descendend strains A/swine/Wisconsin/1/1957 and A/swine/Indiana/A00968365/2012.</li>
<li><em>selectedsites_humanH3N2.txt</em> lists the sites in human H3N2 that substituted along the line of descent from A/Aichi/2/1968 to A/Texas/JMM_49/2012 (see <a class="reference external" href="http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#human-h3n2-mutational-trajectory">http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#human-h3n2-mutational-trajectory</a>).</li>
<li><em>selectedsites_swine.txt</em> lists the sites in swine influenza that substituted along the line of descent from A/swine/Wisconsin/1/1957 to A/swine/Indiana/A00968365/2012 (see <a class="reference external" href="http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#swine-influenza-mutational-trajectory">http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#swine-influenza-mutational-trajectory</a>).</li>
<li><em>epistaticsites_humanH3N2.txt</em> lists the sites in human H3N2 that have been experimentally identified as being involved in epistatic interactions in <a class="reference external" href="http://elife.elifesciences.org/content/2/e00631">Gong et al, 2013</a>. Specifically, these are sites where introducing an individual substitution that actually occurred along the natural evolutionary trajectory is deleterious in the background of the A/Aichi/2/1968 NP.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="running-the-analysis">
<h2>Running the analysis<a class="headerlink" href="#running-the-analysis" title="Permalink to this headline">¶</a></h2>
<p>The entire analysis can be run with the master <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script. Simply run:</p>
<div class="highlight-python"><div class="highlight"><pre>python run_analysis.py
</pre></div>
</div>
<p>This master script creates output <tt class="docutils literal"><span class="pre">*.csv</span></tt> files listing the epitope counts, and also creates plots summarizing the key results. All of the plots are created in <tt class="docutils literal"><span class="pre">*.pdf</span></tt>, <tt class="docutils literal"><span class="pre">*.jpg</span></tt>, and <tt class="docutils literal"><span class="pre">*.eps</span></tt> formats.</p>
<p>Note that the path to the <a class="reference external" href="http://www.drive5.com/muscle/">MUSCLE</a> executable <strong>is hardcoded</strong> in the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script. So if you want to run the script on your computer, you will need to manually edit <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> to specify the appropriate path to <a class="reference external" href="http://www.drive5.com/muscle/">MUSCLE</a>.</p>
</div>
<div class="section" id="steps-in-the-analysis">
<h2>Steps in the analysis<a class="headerlink" href="#steps-in-the-analysis" title="Permalink to this headline">¶</a></h2>
<p>The analysis performed by the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script proceeds through the following steps. These steps are automated by the script, but are documented here individually to explain the overall analysis process.</p>
<div class="section" id="identify-mhc-class-i-epitopes">
<h3>Identify MHC class I epitopes<a class="headerlink" href="#identify-mhc-class-i-epitopes" title="Permalink to this headline">¶</a></h3>
<p>The first step is to identify all MHC class I epitopes for the human H3N2 and swine sequence sets.</p>
<p>This is done by searching for all epitopes between length 8 and 12 residues from non-redundant MHC class I alleles that align with no more than one mismatch to one of the sequences specified in <em>targetprots_humanH3N2.fasta</em> (for human H3N2) or <em>targetprots_swine.fasta</em> (for swine).</p>
<p>The alleles are classified at the level of the allele group (as defined by <a class="reference external" href="http://hla.alleles.org/nomenclature/naming.html">HLA nomenclature</a>), so for example <cite>HLA-B*35:01</cite> has a gene of <cite>HLA-B</cite> and group of <cite>35</cite>, and would be non-redundant with <cite>HLA-B*27:01</cite> but redundant with <cite>HLA-B*35:03</cite>.</p>
<p>The identification of epitopes was done using the <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> package with the following commands:</p>
<div class="highlight-python"><div class="highlight"><pre>epitopefinder_getepitopes.py epitopefinder_getepitopes_human_H3N2_infile.txt
epitopefinder_getepitopes.py epitopefinder_getepitopes_swine_infile.txt
</pre></div>
</div>
<p>where the input files are created automatically by the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script.</p>
<p>The identified epitopes are found in the following created files:</p>
<blockquote>
<div><ul>
<li><p class="first">For human H3N2:</p>
<blockquote>
<div><ul class="simple">
<li><em>epitopesbysite_human_H3N2.csv</em></li>
<li><em>epitopeslist_human_H3N2.csv</em></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">For swine:</p>
<blockquote>
<div><ul class="simple">
<li><em>epitopesbysite_swine.csv</em></li>
<li><em>epitopeslist_swine.csv</em></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
<p>Plots are then created showing the number of epitopes per site for the swine and human H3N2 NPs. The plots are created using <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> with the commands:</p>
<div class="highlight-python"><div class="highlight"><pre>epitopefinder_plotdistributioncomparison.py epitopefinder_plotdistributioncomparison_human_H3N2_vs_swine_infile.txt
epitopefinder_plotlineardensity epitopefinder_plotlineardensity_infile.txt
</pre></div>
</div>
<p>where the input files are created automatically by the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script.</p>
<p>These plots show that the number and distribution of epitopes among the sites are highly similar for human H3N2 and swine NPs.</p>
<blockquote>
<div><div class="figure align-center">
<a class="reference internal image-reference" href="_images/epitopelineardensity1.jpg"><img alt="epitopelineardensity.jpg" src="_images/epitopelineardensity1.jpg" style="width: 80%;" /></a>
<p class="caption">Above is the <tt class="docutils literal"><span class="pre">epitopelineardensity.*</span></tt> plot.</p>
</div>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/distributioncomparison_human_H3N2_vs_swine.jpg"><img alt="distributioncomparison_human_H3N2_vs_swine.jpg" src="_images/distributioncomparison_human_H3N2_vs_swine.jpg" style="width: 50%;" /></a>
<p class="caption">Above is the <tt class="docutils literal"><span class="pre">distributioncomparison_human_H3N2_vs_swine.*</span></tt> plot.</p>
</div>
</div></blockquote>
<p>The <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script also performs a <a class="reference external" href="http://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov-Smirnov test</a> of the null hypothesis that the distributions of epitopes for the human H3N2 and swine NPs are drawn from the same underlying distribution. This null hypothesis is consistent with the data (P = 1.00), suggesting that the epitopes are equivalently distributed for the two NPs.</p>
</div>
<div class="section" id="analyze-epitopes-among-sites-that-substitute">
<h3>Analyze epitopes among sites that substitute<a class="headerlink" href="#analyze-epitopes-among-sites-that-substitute" title="Permalink to this headline">¶</a></h3>
<p>The next step is to extract information about the number of epitopes per site for the sites of all substitutions along the evolutionary trajectory of interest. Recall that the evolutionary trajectories for the human H3N2 and swine NPs are found at <a class="reference external" href="http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#human-h3n2-mutational-trajectory">http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#human-h3n2-mutational-trajectory</a> and <a class="reference external" href="http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#swine-influenza-mutational-trajectory">http://jbloom.github.io/mutpath/example_influenza_NP_1918_Descended.html#swine-influenza-mutational-trajectory</a>.</p>
<p>The identification of the epitopes among the sites that substitute is done using the <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> package with the following commands:</p>
<div class="highlight-python"><div class="highlight"><pre>epitopefinder_selectsites.py selectsites_infile_human_H3N2.txt
epitopefinder_selectsites.py selectsites_infile_swine.txt
</pre></div>
</div>
<p>where the input files are created automatically by the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script.</p>
<p>The number of epitopes per site for the selected sites are listed in the following created files:</p>
<blockquote>
<div><ul class="simple">
<li><em>epitopecounts_selectedsites_human_H3N2.csv</em></li>
<li><em>epitopecounts_selectedsites_swine.csv</em></li>
</ul>
</div></blockquote>
<p>Plots are then created comparing the number of epitopes per site among all sites versus the subset of sites that substituted along the evolutionar trajectories. The plots are created using <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> with the commands:</p>
<div class="highlight-python"><div class="highlight"><pre>epitopefinder_plotdistributioncomparison.py epitopefinder_plotdistributioncomparison_human_H3N2_all_vs_substituted_infile.txt
epitopefinder_plotdistributioncomparison.py epitopefinder_plotdistributioncomparison_swine_all_vs_substituted_infile.txt
</pre></div>
</div>
<p>where the input files are created automatically by the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script.</p>
<p>The P-values shown on the plots are one-sided P-values giving the probability that by random chance, the subset would have a difference from all sites as great as that observed. These plots indicate that the sites that substitute in swine NP have significantly fewer epitopes than the average site, while for human H3N2 the sites that substitute are bordering on significantly having more epitopes than all sites. One possible interpretation: on average human CTLs target sites that are relatively intolerant to mutation, so in the absence of selection (swine NP) the typical substituted site is depleted in epitopes. But in human H3N2 NP, pressure for CTL escape causes sites that substitute to have an excess of epitopes.</p>
<blockquote>
<div><div class="figure align-center">
<a class="reference internal image-reference" href="_images/distributioncomparison_human_H3N2_all_vs_substituted.jpg"><img alt="distributioncomparison_human_H3N2_all_vs_substituted.jpg" src="_images/distributioncomparison_human_H3N2_all_vs_substituted.jpg" style="width: 50%;" /></a>
<p class="caption">Above is the <tt class="docutils literal"><span class="pre">distributioncomparison_human_H3N2_all_vs_substituted.*</span></tt> plot.</p>
</div>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/distributioncomparison_swine_all_vs_substituted.jpg"><img alt="distributioncomparison_swine_all_vs_substituted.jpg" src="_images/distributioncomparison_swine_all_vs_substituted.jpg" style="width: 50%;" /></a>
<p class="caption">Above is the <tt class="docutils literal"><span class="pre">distributioncomparison_swine_all_vs_substituted.*</span></tt> plot.</p>
</div>
</div></blockquote>
<p>The <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script test the null hypothesis that the average number of epitopes for the substituted sites for human H3N2 do not exceed those for the substituted sites in swine NP more than is by expected by chance. This is done by drawing paiared random samples of the same sizes as the actual sets of substituted sites from the set of all site epitope counts for each NP homolog, and then testing how many paired random subsets have a difference in the human H3N2 and swine subset means that is greater than the actual difference observed. The resulting P-value is 0.008, indicating that the substituted sites for human NP contain significantly more epitopes than for swine NP.</p>
</div>
<div class="section" id="analyze-epitopes-among-epistatic-sites">
<h3>Analyze epitopes among &#8220;epistatic&#8221; sites<a class="headerlink" href="#analyze-epitopes-among-epistatic-sites" title="Permalink to this headline">¶</a></h3>
<p>The next step is to examine epitopes among the &#8220;epistatic&#8221; sites. These are the three sites of substitutions in human H3N2 (residues 259, 280, and 384) that were identified in <a class="reference external" href="http://elife.elifesciences.org/content/2/e00631">Gong et al, 2013</a> as being individually deleterious in the background of the A/Aichi/2/1968 human H3N2 NP. Note that there are no such &#8220;epistatic&#8221; sites along the swine trajectory of interest, so no analysis can be performed of epistatic sites in swine NP.</p>
<p>The identification of the epitopes among the epistatic sites is done using the <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> package with the following commands:</p>
<div class="highlight-python"><div class="highlight"><pre>epitopefinder_selectsites.py epitopefinder_selectsites_human_H3N2_epistaticsites.txt
epitopefinder_selectsites.py selectsites_infile_swine.txt
</pre></div>
</div>
<p>where the input file is created automatically by the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script.</p>
<p>The number of epitopes per site for the epistatics sites is listed in the following created file:</p>
<blockquote>
<div><ul class="simple">
<li><em>epitopecounts_epistaticsites_human_H3N2.csv</em></li>
</ul>
</div></blockquote>
<p>Plots are then created comparing the number of epitopes per site among the epistatic sites with the numbers among all sites and substituted sites. The plots are created using <a class="reference external" href="https://github.com/jbloom/epitopefinder">epitopefinder</a> with the commands:</p>
<div class="highlight-python"><div class="highlight"><pre>epitopefinder_plotdistributioncomparison.py epitopefinder_plotdistributioncomparison_human_H3N2_all_vs_epistatic_infile.txt
epitopefinder_plotdistributioncomparison.py epitopefinder_plotdistributioncomparison_human_H3N2_substituted_vs_epistatic_infile.txt
</pre></div>
</div>
<p>where the input files are created automatically by the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script.</p>
<p>The P-values shown on the plots are one-sided P-values giving the probability that by random chance, the epistatic subset would have a difference from all sites or the substituted as great as that observed. These plots indicate that the epistatic sites are significantly enriched in epitopes relative to all sites and substituted sites in human H3N2.</p>
<blockquote>
<div><div class="figure align-center">
<a class="reference internal image-reference" href="_images/distributioncomparison_human_H3N2_all_vs_epistatic.jpg"><img alt="distributioncomparison_human_H3N2_all_vs_epistatic.jpg" src="_images/distributioncomparison_human_H3N2_all_vs_epistatic.jpg" style="width: 50%;" /></a>
<p class="caption">Above is the <tt class="docutils literal"><span class="pre">distributioncomparison_human_H3N2_all_vs_epistatic.*</span></tt> plot.</p>
</div>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/distributioncomparison_human_H3N2_substituted_vs_epistatic.jpg"><img alt="distributioncomparison_human_H3N2_all_vs_substituted.jpg" src="_images/distributioncomparison_human_H3N2_substituted_vs_epistatic.jpg" style="width: 50%;" /></a>
<p class="caption">Above is the <tt class="docutils literal"><span class="pre">distributioncomparison_human_H3N2_substituted_vs_epistatic.*</span></tt> plot.</p>
</div>
</div></blockquote>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">CTL epitopes in influenza NP: human H3N2 and swine</a><ul>
<li><a class="reference internal" href="#software-used">Software used</a></li>
<li><a class="reference internal" href="#input-data-files">Input data files</a></li>
<li><a class="reference internal" href="#running-the-analysis">Running the analysis</a></li>
<li><a class="reference internal" href="#steps-in-the-analysis">Steps in the analysis</a><ul>
<li><a class="reference internal" href="#identify-mhc-class-i-epitopes">Identify MHC class I epitopes</a></li>
<li><a class="reference internal" href="#analyze-epitopes-among-sites-that-substitute">Analyze epitopes among sites that substitute</a></li>
<li><a class="reference internal" href="#analyze-epitopes-among-epistatic-sites">Analyze epitopes among &#8220;epistatic&#8221; sites</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="examples.html"
                        title="previous chapter">Examples</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pythonapi.html"
                        title="next chapter">Python modules</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/example_NP_CTL_epitopes_H3N2_and_swine.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pythonapi.html" title="Python modules"
             >next</a> |</li>
        <li class="right" >
          <a href="examples.html" title="Examples"
             >previous</a> |</li>
        <li><a href="index.html">epitopefinder 0.1 documentation</a> &raquo;</li>
          <li><a href="examples.html" >Examples</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Jesse D. Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.1.
    </div>
  </body>
</html>